singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,250 @@
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.thrift.TException;
+import org.apache.thrift.TSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class AuditEventKafkaSender implements LoggingAuditEventSender {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuditEventKafkaSender.class);
+
+  private final LoggingAuditStage stage;
+  private final String host;
+  private final LinkedBlockingDeque<LoggingAuditEvent> queue;
+  private KafkaProducer<byte[], byte[]> kafkaProducer;
+  private TSerializer serializer = new TSerializer();
+  private AtomicBoolean cancelled = new AtomicBoolean(false);
+  private String topic;
+  private String name;
+  private Thread thread;
+  private int dequeueWaitInSeconds = 30;
+  private int stopGracePeriodInSeconds = 300;
+  private int callingThreadSleepInSeconds = 30;
+  private int numOfPartitionsToTrySending = 3;
+  private Map<Integer, AtomicInteger> partitionErrorCounts = new ConcurrentHashMap<>();
+  private Map<LoggingAuditHeaders, Set<Integer>> eventBadPartitionsMap = new ConcurrentHashMap<>();
+
+  public AuditEventKafkaSender(KafkaSenderConfig config,
+                               LinkedBlockingDeque<LoggingAuditEvent> queue,
+                               LoggingAuditStage stage , String host, String name){
+     this.topic = config.getTopic();
+     this.queue = queue;
+     this.stage = stage;
+     this.host = host;
+     this.name = name;
+  }
+
+  public KafkaProducer<byte[], byte[]> getKafkaProducer() {
+    return kafkaProducer;
+  }
+
+  public void setKafkaProducer(KafkaProducer<byte[], byte[]> kafkaProducer) {
+    this.kafkaProducer = kafkaProducer;
+  }
+
+
+  public int getGoodPartition(Set<Integer> badPartitions, int numOfPartitions){
+    int randomPartition = 0;
+    int trials = 10;
+    while(trials > 0){
+      trials -= 1;
+      randomPartition = ThreadLocalRandom.current().nextInt(numOfPartitions);
+      if (!badPartitions.contains(randomPartition)){
+        break;
+      }
+    }
+    return randomPartition;
+  }
+
+  @Override
+  public void run() {
+    LoggingAuditEvent event = null;
+    ProducerRecord<byte[], byte[]> record;
+    byte[] value = null;
+    List<PartitionInfo> partitionInfoList = this.kafkaProducer.partitionsFor(topic);
+    LOG.info("Partitions of {}: {}", topic, partitionInfoList);
+    for(PartitionInfo partitionInfo : partitionInfoList){
+      partitionErrorCounts.put(partitionInfo.partition(), new AtomicInteger(0));
+    }
+     while(!cancelled.get()){
```

Comment:
Any reason for still using a loop control?

Reply:
if exception happens,  we catch the exception and break the loop.